File: OneDigitDifferenceNumbers/OneDigitDifferenceNumbers.js

const prompt = require('prompt-sync')();
/**
 * isPrime function
 * @param {number} num - The number to check for primality
 * @returns {boolean} - Indicates whether the number is prime or not
 * @description Function to check if the number is prime or not
 */
function isPrime(number) {
    let num = parseInt(number);
    if (num === 1) return false; // 1 is not a prime number
    for (let i = 2; i <= Math.sqrt(num); i++) {
        if (num % i === 0) return false; // If the number is divisible by any other number, it is not prime
    }
    return true; // The number is prime if it passes all the checks
}

/**
 * countOneDigitDifferenceNumber function
 * @param {string} currentNumber - The current number being evaluated
 * @param {object} answer - An object to store the count of one-digit-difference numbers
 * @param {number} range - The maximum range of numbers to consider
 * @description Recursive function to find the number of one-digit-difference numbers
 */
function countOneDigitDifferenceNumber(currentNumber, answer, range) {
    if (currentNumber.length > range) 
        return; // Stop recursion if the number length exceeds the range
    if (isPrime(currentNumber)) 
        answer.count++; // if the last digit is 0, then there is only one next possible digit i.e 1
    if (currentNumber[currentNumber.length - 1] === '0') {
        // if the last digit is 0, then there is only one next possible digit i.e 1
        countOneDigitDifferenceNumber(currentNumber + '1', answer, range);
    } else if (currentNumber[currentNumber.length - 1] === '9') {
        // if the last digit is 9, then there is only one next possible digit i.e 8
        countOneDigitDifferenceNumber(currentNumber + '8', answer, range);
    } else {
        // else there are two possible digits i.e currentNumber+1 and currentNumber-1
        const lastDigit = parseInt(currentNumber[currentNumber.length - 1]);
        countOneDigitDifferenceNumber(currentNumber + (lastDigit + 1), answer, range);
        countOneDigitDifferenceNumber(currentNumber + (lastDigit - 1), answer, range);
    }
}
/**
 * validateInput function
 * @param {number} input - The input to be validated
 * @returns {boolean} - Indicates whether the input is valid or not
 * @description Function to validate the input
 */
module.exports.validateInput = (input) => {
    /*
    * The input should be an integer
    * The input should be greater than 0
    * The input should be less than 30
    * 
    */
    if (!Number.isInteger(input) || input < 0 || input > 30) {
        return false;
    }
    return true;
}

/**
 * solution function
 * @param {number} range - The maximum range of numbers to consider
 * @returns {number} - The number of one-digit-difference numbers
 * @description Function to find the number of one-digit-difference numbers
 * 
 */
// module.exports.solution = async (range) => {
//     let answer = {count : 0};
//     return new Promise((resolve, reject) => {
//         for (let i = 1; i < 10; i++) {
//             countOneDigitDifferenceNumber(i.toString(), answer, range);
//           }
//         console.log(answer.count);
//         resolve(answer.count);
//     });
// }
module.exports.solution = async (range) => {
    let answer = { count: 0 };
    return new Promise((resolve, reject) => {
        for (let i = 1; i < 10; i++) {
            countOneDigitDifferenceNumber(i.toString(), answer, range);
        }
        console.log(answer.count);
        resolve(answer.count);
    });
};


 
  

File: OneDigitDifferenceNumbers/OneDigitDifferenceNumbers.cpp
/**
 * @file Odd.cpp
 * @brief This file contains the code to find the number of one-digit-difference numbers under 10^n
 * @details The code is written in C++ language. The code is tested on Windows 10 with Visual Studio 2019.
 * @version 1.0
 * @date 2021-09-30
 * @author Shaik Riyaz
*/

/**
 * @file Odd.cpp
 * @brief This file contains the code to find the number of one-digit-difference numbers under 10^n
 * @details The code is written in C++ language. The code is tested on Windows 10 with Visual Studio 2019.
 * @version 1.0
 * @date 2021-09-30
 * @author Shaik Riyaz
*/


#include<iostream> // for input and output
#include<string> // for string operations
#include<chrono> // for calculating execution time
using namespace std;
/**
 * isPrime function 
 * @param num: string
 * @return bool
 * @description: function to check if the number is prime or not
 */
bool isPrime(string num) {
    // convert the string to long long as it is easier to check for prime and also to avoid overflow
    int n = stoll(num);

    // if the number is 1, then it is not prime
    if (n == 1) return false;

    // it is enough to check till sqrt(n) to check if the number is prime or not
    //reason: if n is not prime, then it can be written as n = a*b and either a or b is less than sqrt(n)
    for (int i = 2;i * i <= n;i++) {
        if (n % i == 0) 
            return false;
    }
    // number is not divisible by any number, so it is prime
    return true;
}


/**
 * oddNumbers function
 * @param currentNumber: string
 * @param count: int
 * @param range: int
 * @return void
 * @description recursive function to find the number of one-digit-difference numbers
 */
void oddNumbers(string currentNumber, int& count, int& range) {
   // base case if the num is greater than the 10 ^ range then stop the recursion
    if (currentNumber.size() > range)
        return;
    // if the number is prime, increment the count
    if (isPrime(currentNumber))
        count++;
    // if the last digit is 0, then there is only one next possible digit i.e 1
    if (currentNumber[currentNumber.size() - 1] == '0') {
        oddNumbers(currentNumber + '1', count, range);
    }
    // if the last digit is 9, then there is only one next possible digit i.e 8
    else if (currentNumber[currentNumber.size() - 1] == '9')
        oddNumbers(currentNumber + '8', count, range);
    // else there are two possible digits i.e n+1 and n-1
    else {
        // get the last digit
        int n = currentNumber[currentNumber.size() - 1] - '0';
        oddNumbers(currentNumber + to_string((n + 1)), count, range);
        oddNumbers(currentNumber + to_string((n - 1)), count, range);
    }
}
/**
 * main function
 * @description: main function to join all the modules.
 * @return 0
*/
int main() {
    
    // record the start time
    int count = 0, range;
    cin>>range;
    if(range <= 0)
        cout<<"Input is invalid try again."<<endl;
    auto start = chrono::high_resolution_clock::now();
    for (int i = 1; i < 10; i++) {
        oddNumbers(to_string(i), count, range);
    }
    // record the end time
    auto end = chrono::high_resolution_clock::now();
    cout << "No of one-digit-difference numbers under 10^" << range<<" are:" << count << endl;
    // calculate the execution time
    auto executionTime = chrono::duration_cast<std::chrono::milliseconds>(end - start);
    cout << "The execution time of the function is " << executionTime.count() << " milliseconds.\n";
}